### PR TITLE
Cloudwatch Metrics: Adjust error handling

### PR DESCRIPTION
**Pull request by idastambuk**
**What is this feature?**
Partly fixes [this](https://github.com/grafana/grafana/issues/78819): 

> The only way I see the error is in the network tab of my browser which is not a great experience as a user, as it seems to imply it was a valid query there just was not data associated:

The problem here was that we had a catchError block in MetricsRunner that was getting skipped because we're using DataSourceWithBackend's `query` method now. There, the response from Cloudwatch BE is getting processed before it goes into the Metrics Runner. The way errors are processed is that the response is [turned into an error object and transformed into a observable](https://github.com/grafana/grafana/blob/3c4bc24deca0008d7d0e57daddf2df4f77faab04/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L254): 
```
catchError((err) => {
          return of(toDataQueryResponse(err));
        })
```
This response doesn't get caught in the `catch` block of the CWMetricsQueryRunner, so the errors weren't being propagated to the panel. 

The biggest change here is that I removed toast alerts for errors in Cloudwatch. These errors could be disruptive to the user, as we display an error icon with every panel that it anyway: 

Screenshot 2023-12-28 at 10 50 18

I looked for the throwError usage elsewhere in grafana and it's used in Loki, Prometheus and Graphite, but AFAIK in none of the AWS datasources. 

However, throttling errors were handled separately before, i.e. errors concerning CW limits: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html
I kept this handling, just added "Rate exceeded"  (https://aws.amazon.com/blogs/mt/managing-monitoring-api-throttling-in-workloads/) Im not sure if this should be removed in favor of just displaying the errors in the top left corner of the panel. But it could be argued that if a user has a dashboard with many CW panels and they would want to know if they hit any quotas immediately upon opening the dashboard.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**
This should be done for Logs queries as well, since they're broken too, but that will be done in another PR

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/tsdb/cloudwatch/time_series_query.go [hunk 1]
```diff
@@ -121,6 +122,7 @@ func (e *cloudWatchExecutor) executeTimeSeriesQuery(ctx context.Context, logger
 			Error: fmt.Errorf("metric request error: %q", err),
 		}
 		resultChan <- &responseWrapper{
+			RefId:        getQueryRefIdFromErrorString(err.Error(), requestQueries),
```
Comment (idastambuk) [inline pkg/tsdb/cloudwatch/time_series_query.go:125]: This is more of an improvement. We dont really need the refId in the error in order to display the error in the panel top left corner. However, we do need it to display it in the query editor inside the panel.

**Review by iwysiu — COMMENTED** [review]
I tried it and it worked! a couple questions about the tests.

**Review by sarahzinger — CHANGES_REQUESTED** [review]
Hmm tested this locally and I don't think it works? When I run an incorrect query like "nonsense + 2" (that was linked in the issue) I get this weird flash of an error message but then it disappears? Let me know if I'm missing something! 

https://github.com/grafana/grafana/assets/6620164/43e8a7ee-fb4f-46e9-93bc-b3ee8fc7d364

File: public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.test.ts [hunk 2]
```diff
@@ -169,110 +189,111 @@ describe('CloudWatchMetricsQueryRunner', () => {
...
         await expect(runner.handleMetricQueries(queries, request, queryMock)).toEmitValuesWith((received) => {
           const result = received[0];
-          expect(getFrameDisplayName(result.data[0])).toBe(
-            data.results.A.series?.length && data.results.A.series[0].target
-          );
-          expect(result.data[0].fields[1].values[0]).toBe(
-            data.results.A.series?.length && data.results.A.series[0].datapoints[0][0]
-          );
+          expect(getFrameDisplayName(result.data[0])).toBe(1 && 'CPUUtilization_Average');
+          expect(result.data[0].fields[1].values[0]).toBe(1 && 1);
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.test.ts:200]: Oops, it's "1", fixed it!

File: public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.ts [hunk 3]
```diff
@@ -109,12 +110,7 @@ export class CloudWatchMetricsQueryRunner extends CloudWatchRequest {
   ): Observable<DataQueryResponse> {
     return queryFn(request).pipe(
       map((res) => {
-        const dataframes: DataFrame[] = res.data;
-        if (!dataframes || dataframes.length <= 0) {
-          return { data: [] };
-        }
-
-        const lastError = findLast(res.data, (v) => !!v.error);
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.ts:117]: I assume it was in order to only get one error to display. I guess at some point before errors were passed in the res.data, but they're not anymore - they're passed as a separate `errors` array from the DatasourceWithBackend srv, in `response.errors`. As far as I can see there isn't really a way for one metric query to return multiple errors from AWS and our BE

File: public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.ts [hunk 4]
```diff
@@ -125,43 +121,41 @@ export class CloudWatchMetricsQueryRunner extends CloudWatchRequest {
...
+            )
+          ) {
+            const failedRefIds = res.errors.map((error) => error.refId).filter((refId) => refId);
+            if (failedRefIds.length > 0) {
+              const regionsAffected = Object.values(request.targets).reduce(
+                (res: string[], { refId, region }) =>
+                  (refId && !failedRefIds.includes(refId)) || res.includes(region) ? res : [...res, region],
+                []
+              );
+              regionsAffected.forEach((region) => {
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.ts:137]: Yea, regions were only checked for throttling in the old error handling, so I kept it. Regarding other errors, I have some of the reasoning about this is in the description. TBH I originally wanted to remove all toaster alerts and just depend on the regular grafana error ui, so it's definitely something we can discuss

File: public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.ts [hunk 5]
```diff
@@ -125,43 +121,41 @@ export class CloudWatchMetricsQueryRunner extends CloudWatchRequest {
...
         };
       }),
-      catchError((err: DataQueryError<CloudWatchMetricsQuery>) => {
-        const isFrameError = err.data?.results;
-
-        // Error is not frame specific
-        if (!isFrameError && err.data && err.data.message === 'Metric request error' && err.data.error) {
-          err.message = err.data.error;
-          return throwError(() => err);
+      catchError((err: unknown) => {
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.ts:153]: There's a bit more info in the description about what caused this, basically cause DatasourceWithBackend that we use now, returns data and not error

File: public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.test.ts [hunk 6]
```diff
@@ -169,110 +189,111 @@ describe('CloudWatchMetricsQueryRunner', () => {
...
+        },
+      };
 
-        beforeEach(() => {
-          redux.setStore({
-            ...redux.store,
-            dispatch: jest.fn(),
-          });
+      beforeEach(() => {
+        redux.setStore({
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.test.ts:279]: We have a ticket to remove this here: https://github.com/grafana/grafana/issues/80151

File: public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.test.ts [hunk 7]
```diff
@@ -649,58 +707,6 @@ describe('CloudWatchMetricsQueryRunner', () => {
     });
   });
 
-  describe('debouncedCustomAlert', () => {
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchMetricsQueryRunner.test.ts:652]: Accident, I put it back, good catch

**Comment by idastambuk** [comment]
> Hmm tested this locally and I don't think it works? When I run an incorrect query like "nonsense + 2" (that was linked in the issue) I get this weird flash of an error message but then it disappears? Let me know if I'm missing something!
> 
>  Screen.Recording.2024-01-05.at.10.08.32.AM.mov

This PR was probably missing from the branch: https://github.com/grafana/grafana/pull/79943, just merged main into it. Can you pull and try again?

**Review by sarahzinger — APPROVED** [review]
Looks like there's a lint error to fix, but I manually tested and it works for me!

**Pull request merged**